### PR TITLE
New Resource: `azurerm_management_group`

### PR DESCRIPTION
**Pull request by justinbarias**
- [x] Updated documentation
- [x] Written test cases
- [x] Create resource azurerm_management_group
- [ ] Modify azurerm_policy_definition to support management group-level policies

I don't see the need for a data source, as the most you need to interpolate is the ID (which you can do from the resource node).

Nested management groups aren't supported for now.

Also, i was trying to do some validation for the subscription ids (a subscription can only be contained in a single management group) - wasn't sure how to extend the validation engine.. Is it as simple as returning a function callback with type  schema.SchemaValidateFunc?

Lastly, as this is my first time contributing publicly with Go, I wasn't sure if I did the right thing with govendor (pulled in a specific package from the Azure GO SDK).

### CODE REVIEW

File: azurerm/resource_arm_management_group.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	managementGroupProperties := managementgroups.CreateManagementGroupProperties{
+		TenantID:    &armTenantID,
+		DisplayName: &name,
+		Details:     &details,
+	}
+
+	createManagementGroupRequest := managementgroups.CreateManagementGroupRequest{
+		ID:   nil,
+		Type: utils.String("/providers/Microsoft.Management/managementGroups"),
+		Name: &name,
```
[the remark this answers is not in the document]
Comment (justinbarias) [reply azurerm/resource_arm_management_group.go]: tried it out.. gave me a Original Error: autorest/azure: Service returned an error. Status=400 Code="BadRequest" Message="Name of the group doesn't match the group id." error..

File: azurerm/resource_arm_management_group.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+	getResp, getErr := client.Get(ctx, name, "", &recurse, "", "no-cache")
+	if getErr != nil {
+		log.Printf("[DEBUG] Error retrieving management group details")
+		return err
+	}
+	d.SetId(*getResp.ID)
+
+	for _, subscription := range subscriptionIds {
+		data := subscription.(string)
+		log.Printf("[DEBUG] Adding subscriptionId %q to management group %q", data, name)
```
[the remark this answers is not in the document]
Comment (justinbarias) [reply azurerm/resource_arm_management_group.go]: based on the SDK package /mgmt/2018-03-01-preview/management, there are two different clients for managementgroup (client) and subscriptions (subscriptionclient) - in this case though, subscriptionclient really only gets used in the managementgroup context, and purely to associate a subscription to a management group (and not for creating subscriptions).. i think there's merit creating a dedicated subscription object, but for the creation/deletion of subscriptions (which is a different SDK package). My two cents..

**Comment by justinbarias** [comment]
@tombuildsstuff  some updates - tried assigning a policy definition to the management group.. 
I'm getting a 
```
Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned anerror. Status=400 Code="InvalidCreatePolicyAssignmentRequest" Message="The policy definition specified in policy assignment 'policyassignment' is out of scope. Policy definitions should be specified only at or above the policy assignment scope."
```
 error.

Seems changes are needed to the azurerm_policy_definition resource. The policyDefinitions package has a separate method for creating policies to be assigned on the management group level (https://docs.microsoft.com/en-us/rest/api/resources/policydefinitions/createorupdateatmanagementgroup).

I'm thinking adding a parameter for the azurerm_policy_definition resource of "policy_scope" which accepts either "subscription", or "managementgroup".

Thoughts?

File: azurerm/resource_arm_management_group_test.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+`, ri)
+}
+
+func testAzureRmManagementGroup_withSubscriptions(ri int, subscriptionID string) string {
+
+	return fmt.Sprintf(`
+resource "azurerm_management_group" "test" {
+  name         = "acctestmg-%d"
+  subscription_ids = [
+	  "%q"
```
[the remark this answers is not in the document]
Comment (justinbarias) [reply azurerm/resource_arm_management_group_test.go]: Fixed with latest commit.

File: azurerm/config.go [hunk 4]
```diff
@@ -1039,6 +1046,16 @@ func (c *ArmClient) registerPolicyClients(endpoint, subscriptionId string, auth
 	c.policyDefinitionsClient = policyDefinitionsClient
 }
 
+func (c *ArmClient) registerManagementGroupClients(endpoint string, auth autorest.Authorizer) {
```
[the remark this answers is not in the document]
Comment (justinbarias) [reply azurerm/config.go]: Added Microsoft.Management - though pretty sure that provider comes registered by default out of the box.

File: azurerm/resource_arm_management_group_test.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+				Config: testAzureRmManagementGroup_basic(ri),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMManagementGroupExists(resourceName),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAzureRMManagementGroup_withSubscriptions(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (justinbarias) [reply azurerm/resource_arm_management_group_test.go]: Done with latest commit

File: azurerm/config.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/config.go]: _probably_ however folks can opt to unregister them, so we're tending to be cautious here

**Comment by tombuildsstuff** [comment]
hey @justinbarias 

Thanks for pushing those changes - I hope you don't mind but I've taken a look through and have updated the schema/added some extra tests so that we can get this merged :)

> I'm thinking adding a parameter for the azurerm_policy_definition resource of "policy_scope" which accepts either "subscription", or "managementgroup".

Rather than hold this PR on support for that - would you mind opening a separate issue to track this? The idea sounds sensible enough - but I'm thinking this'd be better to do conditionally on a new `management_group_id` field being `!= ""` - what do you think?

Thanks!

**Comment by tombuildsstuff** [comment]
Tests pass:

```
 $ acctests azurerm TestAccAzureRMManagementGroup_

=== RUN   TestAccAzureRMManagementGroup_basic
--- PASS: TestAccAzureRMManagementGroup_basic (56.52s)
=== RUN   TestAccAzureRMManagementGroup_nested
--- PASS: TestAccAzureRMManagementGroup_nested (86.19s)
=== RUN   TestAccAzureRMManagementGroup_multiLevel
--- PASS: TestAccAzureRMManagementGroup_multiLevel (122.92s)
=== RUN   TestAccAzureRMManagementGroup_multiLevelUpdated
--- PASS: TestAccAzureRMManagementGroup_multiLevelUpdated (200.95s)
=== RUN   TestAccAzureRMManagementGroup_withName
--- PASS: TestAccAzureRMManagementGroup_withName (68.17s)
=== RUN   TestAccAzureRMManagementGroup_updateName
--- PASS: TestAccAzureRMManagementGroup_updateName (98.42s)
=== RUN   TestAccAzureRMManagementGroup_withSubscriptions
--- PASS: TestAccAzureRMManagementGroup_withSubscriptions (189.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	823.213s
```
039373e

**Pull request merged**